**The data used in the Notebook was taken from the Medellin Metro open data website.** https://datosabiertos-metrodemedellin.opendata.arcgis.com/documents/569c4b4c1ad54c3da95aa5f195637db2/about

# EDA METRO MEDELLIN

In [3]:
#IMPORT NEEDED LIBRARIES
import pandas as pd

In [4]:
# IMPORT THE DATA INTO THE STRUCTURE OF A DATAFRAME
df_affluence = pd.read_csv('data/Afluencia_2023.csv')
df_affluence.head(5)

Día Línea de Servicio Hora de operación Unnamed: 3 Unnamed: 4  \
0         NaN               NaN              4:00       5:00       6:00   
1  01/01/2023           LÍNEA 1              192        811        896    
2  01/01/2023           LÍNEA 2               11        123        116    
3  01/01/2023           LÍNEA A         1,328.00   5,104.00   5,701.00    
4  01/01/2024           LÍNEA B           221.00     701.00     750.00    

  Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 13  \
0       7:00       8:00       9:00      10:00      11:00  ...       15:00   
1       716        673        922      1,023      1,208   ...      1,629    
2        95        102        144        147        189   ...        191    
3  4,309.00      4,054      4,852      5,880      7,629   ...     10,999    
4    655.00        628        863      1,085      1,228   ...      1,930    

  Unnamed: 14 Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19  \
0       16:00       17:00       18:00       19:00       20:00       21:00   
1      1,985       2,361       2,694       2,303       1,839       1,478    
2        255         280         317         252         164         111    
3     13,027      15,965      19,123      15,663      12,918       9,509    
4      2,247       2,498       2,648       2,029       1,472       1,087    

  Unnamed: 20 Unnamed: 21                          Unnamed: 22  
0       22:00       23:00  Total general (Número de pasajeros)  
1        236          NaN                              25,483   
2          7          NaN                               3,210   
3        819          NaN                             166,047   
4        151          NaN                              25,106   

[5 rows x 23 columns]

## Checking headers

In [6]:
df_affluence.shape

(4304, 23)

- There are 4,304 rows and 23 columns. We need to delete the first row and rename the name of all columns.

In [8]:
#Renaming columns to English
df_affluence= df_affluence.rename(columns={
    "Día":"Day",
    "Línea de Servicio": "Service_Line",
    "Hora de operación": "04:00",
    "Unnamed: 3": "05:00",
    "Unnamed: 4": "06:00",
    "Unnamed: 5": "07:00",
    "Unnamed: 6": "08:00",
    "Unnamed: 7": "09:00",
    "Unnamed: 8": "10:00",
    "Unnamed: 9": "11:00",
    "Unnamed: 10": "12:00",
    "Unnamed: 11": "13:00",
    "Unnamed: 12": "14:00",
    "Unnamed: 13": "15:00",
    "Unnamed: 14": "16:00",
    "Unnamed: 15": "17:00",
    "Unnamed: 16": "18:00",
    "Unnamed: 17": "19:00",
    "Unnamed: 18": "20:00",
    "Unnamed: 19": "21:00",
    "Unnamed: 20": "22:00",
    "Unnamed: 21": "23:00",
    "Unnamed: 22": "Grand Total (Number of passengers)"    
})

df_affluence = df_affluence.drop(df_affluence.index[0]) #Deleting the first row

## Checking Null or NA Values

In [10]:
df_affluence.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4303 entries, 1 to 4303
Data columns (total 23 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Day                                 4303 non-null   object
 1   Service_Line                        4303 non-null   object
 2   04:00                               3638 non-null   object
 3   05:00                               3643 non-null   object
 4   06:00                               3647 non-null   object
 5   07:00                               3646 non-null   object
 6   08:00                               4092 non-null   object
 7   09:00                               4290 non-null   object
 8   10:00                               4289 non-null   object
 9   11:00                               4289 non-null   object
 10  12:00                               4293 non-null   object
 11  13:00                               4294 non-null   obje

- All NaN values must be replaced with 0.

In [12]:
df_affluence.fillna('0',inplace = True) #replacing missing values with str 0

## Checking Datatypes

In [14]:
df_affluence.dtypes

Day                                   object
Service_Line                          object
04:00                                 object
05:00                                 object
06:00                                 object
07:00                                 object
08:00                                 object
09:00                                 object
10:00                                 object
11:00                                 object
12:00                                 object
13:00                                 object
14:00                                 object
15:00                                 object
16:00                                 object
17:00                                 object
18:00                                 object
19:00                                 object
20:00                                 object
21:00                                 object
22:00                                 object
23:00                                 object
Grand Tota

- "Day" column should be Date datatype (DDMMYYYY)
- Column's index from 2 to 22 should be int64 datatype


In [16]:
# *Changing "Day" column datatype
from datetime import datetime
df_affluence["Day"] = pd.to_datetime(df_affluence["Day"], format='%d/%m/%Y')
df_affluence.dtypes

# *Changing datatype from columns 2 to 22.
objectdt_columns = list(df_affluence.select_dtypes(include=['object']).columns[1:]) #creating a list with all the columns with dtype = object, except for the first one
              
def change_datatype(row):
    casted_int = 0
    if isinstance(row,str):
        row = row.replace(",","").strip().replace(".00","") #the commas inside the columns must be deleted.
    try:
        casted_int = int(row)
    except Exception as err:
        print(f'error: {err}')
    
    return casted_int

df_affluence[objectdt_columns] = df_affluence[objectdt_columns].map(change_datatype)

In [17]:
df_affluence.info() #Checking

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4303 entries, 1 to 4303
Data columns (total 23 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Day                                 4303 non-null   datetime64[ns]
 1   Service_Line                        4303 non-null   object        
 2   04:00                               4303 non-null   int64         
 3   05:00                               4303 non-null   int64         
 4   06:00                               4303 non-null   int64         
 5   07:00                               4303 non-null   int64         
 6   08:00                               4303 non-null   int64         
 7   09:00                               4303 non-null   int64         
 8   10:00                               4303 non-null   int64         
 9   11:00                               4303 non-null   int64         
 10  12:00                   

## Checking rows

Lets check the rows on "Day" and "Service_Line" columns:

In [20]:
#Checking the date range of the data ("Day" column)
unique_years = df_affluence['Day'].dt.year.unique()
unique_months = df_affluence['Day'].dt.month.unique()
unique_days = df_affluence['Day'].dt.day.unique()
print(f'years:{unique_years}\n months: {unique_months}\n days: {unique_days}')

years:[2023 2024]
 months: [ 1  2  3  4  5  6  7  8  9 10 11 12]
 days: [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31]


In [21]:
list(df_affluence["Service_Line"].unique())

['LÍNEA 1',
 'LÍNEA 2',
 'LÍNEA A',
 'LÍNEA B',
 'LÍNEA H',
 'LÍNEA J',
 'LÍNEA K',
 'LÍNEA L',
 'LÍNEA M',
 'LÍNEA O',
 'LÍNEA P',
 'LÍNEA T-A']

- In column "Day" there is a date from 2024. We choose to remove it, since there should be only data from 2023.
- In column "Service_Line" there is an atypical line 'LÍNEA T-A'. After further investigation, line 'LÍNEA T-A' is actually 'LÍNEA T', so we need to replace this value on all rows.

In [23]:
year_2024 = 2024

rows_2024 = df_affluence[df_affluence['Day'].dt.year == 2024] #Checking which rows have 2024 data.

rows_2024.head()

Day Service_Line  04:00  05:00  06:00  07:00  08:00  09:00  10:00  \
4 2024-01-01      LÍNEA B    221    701    750    655    628    863   1085   

   11:00  ...  15:00  16:00  17:00  18:00  19:00  20:00  21:00  22:00  23:00  \
4   1228  ...   1930   2247   2498   2648   2029   1472   1087    151      0   

   Grand Total (Number of passengers)  
4                               25106  

[1 rows x 23 columns]

There is only 1 row with 2024 data, located on index 4

In [25]:
df_affluence.drop(index=4, inplace= True) #deleting the row
df_affluence.reset_index() #We need to reset the index since we cleaned the rows.

index        Day Service_Line  04:00  05:00  06:00  07:00  08:00  09:00  \
0         1 2023-01-01      LÍNEA 1    192    811    896    716    673    922   
1         2 2023-01-01      LÍNEA 2     11    123    116     95    102    144   
2         3 2023-01-01      LÍNEA A   1328   5104   5701   4309   4054   4852   
3         5 2023-01-01      LÍNEA H      0      0      0      0      0     56   
4         6 2023-01-01      LÍNEA J      0      0      0      0      0    470   
...     ...        ...          ...    ...    ...    ...    ...    ...    ...   
4297   4299 2023-12-31      LÍNEA L      0      0      0      0     22     88   
4298   4300 2023-12-31      LÍNEA M      0      0      0      1     12    296   
4299   4301 2023-12-31      LÍNEA O     25    314    236    237    224    319   
4300   4302 2023-12-31      LÍNEA P      0      0      0      0     76    696   
4301   4303 2023-12-31    LÍNEA T-A    135    595    711    723   1049   1207   

      10:00  ...  15:00  16:00  17:00  18:00  19:00  20:00  21:00  22:00  \
0      1023  ...   1629   1985   2361   2694   2303   1839   1478    236   
1       147  ...    191    255    280    317    252    164    111      7   
2      5880  ...  10999  13027  15965  19123  15663  12918   9509    819   
3        62  ...     55    128     99    109     68     46     27      1   
4       410  ...    688    710    857    933    651    311    146      3   
...     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
4297    204  ...    273    254     92      0      0      0      0      0   
4298    295  ...    420    413    400    587    511    431    100      0   
4299    282  ...    279    300    412    477    339    168     38      0   
4300    693  ...    590    700    625    659    634    662     61      0   
4301   1277  ...   1783   1899   1943   2308   1917   1613    328      0   

      23:00  Grand Total (Number of passengers)  
0         0                               25483  
1         0                                3210  
2         0                              166047  
3         0                                 970  
4         0                                7209  
...     ...                                 ...  
4297      0                                2389  
4298      0                                4873  
4299      0                                4758  
4300      0                                7768  
4301      0                               23961  

[4302 rows x 24 columns]

In [26]:
#Replacing 'LÍNEA T-A' for'LÍNEA T'
df_affluence['Service_Line'] = df_affluence['Service_Line'].str.replace('LÍNEA T-A','LÍNEA T')

Now lets see if there are any repetead rows

In [28]:
# Deleting repeated rows
print(f'Set size before deleting repeated rows: {df_affluence.shape}')
df_affluence.drop_duplicates(inplace=True)
print(f'Set size before deleting repeated rows: {df_affluence.shape}')

Set size before deleting repeated rows: (4302, 23)
Set size before deleting repeated rows: (4302, 23)


- There are no repeated rows.

## Checking columns

Lets check if there are any irrelevant columns

In [32]:
#Checking irrelevant columns
cat_columns = ['Day','Service_Line']

for col in cat_columns:
  print(f'Column "{col}": {df_affluence[col].nunique()} sublevels')

Column "Day": 365 sublevels
Column "Service_Line": 12 sublevels


- All categorical columns have more than 1 sublevel, so we won't delete any of them. Also, the "Day" column has 365 days, so it covers all 2023 year.

# Reshaping data

The purpose of this is to significantly reduce the number of columns in our data.

In [36]:
#Unpivot rows from 2 to 21, leaving two columns named "time", and "value"
columns_to_pivot = ['04:00', '05:00', '06:00', '07:00', '08:00', '09:00', '10:00', '11:00','12:00', '13:00', '14:00', '15:00', '16:00', '17:00', '18:00', '19:00', '20:00', '21:00', '22:00', '23:00']
columns_to_keep =["Day","Service_Line","Grand Total (Number of passengers)"]
df_reshaped = pd.melt(df_affluence, id_vars=columns_to_keep, value_vars=columns_to_pivot,var_name="Time", value_name="Passengers",ignore_index="False")

In [37]:
df_reshaped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86040 entries, 0 to 86039
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Day                                 86040 non-null  datetime64[ns]
 1   Service_Line                        86040 non-null  object        
 2   Grand Total (Number of passengers)  86040 non-null  int64         
 3   Time                                86040 non-null  object        
 4   Passengers                          86040 non-null  int64         
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 3.3+ MB


In [38]:
#We now need to change "Time" column datatype, and merge columns "Day" and "Time"
df_reshaped["Time"] = pd.to_datetime(df_reshaped["Time"], format='%H:%M').dt.time #Change "Time" column datatype
df_reshaped["Day_and_Time"] =pd.to_datetime(df_reshaped['Day'].astype(str) + ' ' + df_reshaped['Time'].astype(str))


In [39]:
#For blending purposes with another table, we will replace the word "Línea" from "Service_Line" Column to "Line "
df_reshaped['Service_Line'] = df_reshaped['Service_Line'].str.replace('LÍNEA ', 'Line ')


In [40]:
#Finally, we delete the irrelevant columns:
df_reshaped.drop(["Day","Grand Total (Number of passengers)","Time"], axis=1, inplace=True)

We can create a new column that tells us the day of the week, for deeper analysis

In [42]:
df_reshaped['Day_of_the_week'] = df_reshaped['Day_and_Time'].dt.day_name()

# Exporting the DataFrame

Finally, we export the cleaned Dataframe

In [45]:
filename = f'data/Exported_data/Affluence_2023.csv' #path for the cleaned Affluence_2023 csv data
df_reshaped.to_csv(filename, index=False)
print("File has been created succesfully")

File has been created succesfully
